source/features/quick-file-edit.tsx

```diff
@@ -9,24 +9,35 @@ import features from '.';
 import GitHubURL from '../github-helpers/github-url';
 import {isPermalink} from '../github-helpers';
 import getDefaultBranch from '../github-helpers/get-default-branch';
-import onFileListUpdate from '../github-events/on-file-list-update';
-
-async function init(): Promise<void> {
-	const isPermalink_ = await isPermalink();
-	for (const fileIcon of select.all('.js-navigation-container .octicon-file')) {
-		const fileLink = fileIcon.closest('.js-navigation-item')!.querySelector('a.js-navigation-open')!;
-		const url = new GitHubURL(fileLink.href).assign({
-			route: 'edit',
-		});
-
-		if (isPermalink_) {
-			// eslint-disable-next-line no-await-in-loop
-			url.branch = await getDefaultBranch(); // Permalinks can't be edited
-		}
-
-		wrap(fileIcon, <a data-skip-pjax href={url.href} className="rgh-quick-file-edit"/>);
-		fileIcon.after(<PencilIcon/>);
+import observe from '../helpers/selector-observer';
+
+// For https://github.com/refined-github/refined-github/issues/5821
+// TODO: Maybe drop it after `isPermalink` accepts a value thanks to https://github.com/refined-github/github-url-detection/issues/85
+const cachePerPage = {
+	cacheKey: () => location.pathname,
+};
+
+const cachedIsPermalink = mem(isPermalink, cachePerPage);
+const cachedGetDefaultBranch = mem(getDefaultBranch, cachePerPage);
+
+async function linkifyIcon(fileIcon: Element): Promise<void> {
```

Comment:
Tested while going back and forth, changing folder, branches. All solid.

Diff better seen without whitespace

- https://github.com/refined-github/sandbox

<img width="279" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/190650219-49015b2c-8fc8-457d-89a0-ef791c3f6df5.png">